File: find_all_duplicates_in_an_array_442.java
class Solution {
    public List<Integer> findDuplicates(int[] nums) {
        List<Integer> li=new ArrayList<Integer>();
        for(int i=0;i<nums.length;i++){
            int id=Math.abs(nums[i]);
            if(nums[id-1]<0) li.add(id);
            else nums[id-1]=-nums[id-1];

        }
    return li;
    }
}

File: find_all_duplicates_in_an_array_442.py
class Solution:
    def findDuplicates(self, nums: List[int]) -> List[int]:
        ans = []
    # Use a dictionary to count occurrences of each element
        counts = {}
    
        for i in nums:
            if i in counts:
                counts[i] += 1
            else:
                counts[i] = 1
    
    # Append elements that appear exactly twice
        for key, value in counts.items():
            if value == 2:
                ans.append(key)
    
        return ans